#################################################################################
#
#                         African herbivores gut microbiome
#                      
#              Rojas et al 2021.Host phylogeny and host ecology structure
#               the mammalian gut microbiota at different taxonomic scales
#
#                               By: Connie Rojas
#                               Created: 5 Sept 2020
#                            Last updated: 18 Jan 2021
#
################################################################################

##CODE FOR: generating a matrix of evolutionary divergence times between each pair 
#of host species based on phylogenetic trees downloaded from: 
#Upham, N. S., J. A. Esselstyn, and W. Jetz. 2019. Inferring the mammal tree: species-level 
#sets of phylogenies for questions in ecology, evolution, and conservation. PLOS Biology.

#this data will be necessary for running a mantel test to determine whether gut microbiota 
#similarity increases with host phylogenetic relatedness

source(file="scripts/00_background.R"); #load necessary packages and specifications


################################################################################
#             1. Load nexus file generated by vertlife.org (1000 trees; 
#                 DNA-only trees; artiodactyla & elephant)
#             for exact configurations, see data/06_vertlife.config.yaml
################################################################################
trees <- ape::read.nexus("data/06_Upham2019_output.nex");
load("data/01_sample_metadata_filtered.Rdata");  #also load sample metadata


################################################################################
#             2. create the comparison pairs (e.g. Giraffe vs. Elephant)
################################################################################
spdf=meta[!duplicated(meta$species_short),];
spdf[7,6] <- "Tragelaphus_oryx"; ### rename Topi to match Upham 2019
spdf[9,6] <- "Giraffa_camelopardalis"; ### drop the subspp to match Upham 2019
short <- spdf$species_short;
taxa <- spdf$tax_name %>%  gsub(" ","_",.);
taxa_pairs <- combn(taxa,2,simplify=FALSE);


################################################################################
#             3. prune phylogenetic tree to only include study species 
################################################################################
plan(multisession);
master_matrix <- trees %>% furrr::future_map(function(j){
  ages <- taxa_pairs %>% map_dbl(function(i){ 
    j %>% getMRCA(.,tip = i) %>% 
      extract.clade(j,node = .) %>% 
      branching.times(.) %>% max(.,na.rm=T)
  });
  return(ages);
},.progress = TRUE);

################################################################################
#             4. create matrix of pairwise divergence times 
#         these are the means across the 1,000 trees downloaded from VertLife
################################################################################
#create matrix of pairwise divergence times
#
mega_matrix <- master_matrix %>% furrr::future_map(function(x){
  age_matrix <- matrix(nrow = dim(spdf)[1],ncol = dim(spdf)[1],dimnames= list(short,short))
  age_matrix[lower.tri(age_matrix)] <- x
  age_matrix[upper.tri(age_matrix)] <- t(age_matrix)[upper.tri(age_matrix)]
  diag(age_matrix ) <- 0
  return(age_matrix)
},.progress=TRUE);
plan(sequential);

mega_matrix[[1]]
arr <- array( unlist(mega_matrix) , c(11,11,length(mega_matrix)));
divergence_times <- apply( arr , 1:2 , function(x) round(mean(x),5 ));
dimnames(divergence_times) <- list(short,short);

################################################################################
#             5. save matrix of host pairwise divergence times 
################################################################################
save(divergence_times, file="data/06_divergence_times_matrix.Rdata");

#Code for generating a phylogeny of the herbivore species is not available, but
#the actual phylogeny (.pdf) can be found in figures/06_host_phylogeny.pdf