#Manuscript: Host phylogeny and host ecology structure the mammalian gut microbiota 
#at different taxonomic scales
#Rojas et al. 2020
#Code for generating phylogenetic tree of ASVs to eventually calculate Faith's Phylogenetic Diversity (PD)
#AFTER SAMPLES WERE SUBSAMPLED TO 17000

source(file="scripts/background.R") #load necessary packages and specifications

##load ASV table generated by mothur software (Schloss et.al 2009)
#samples were subsampled to 17000 sequences; causing 23 samples to be removed 
masv=read.table("data/mothur.asvtable", sep="\t",header=T);

#clean up data 
masv$label=NULL; masv$numOtus=NULL;
rownames(masv)=masv$Group; masv$Group=NULL;

#replace ASV labels with ASV sequence names
#ASV sequence names are the column names of the ASV table output by DADA2
load("data/ASV_names.Rdata");
colnames(masv)=rownames(ASV)[match(names(masv),ASV$ASV)];
masv=as.matrix(masv);

#build phylogenetic tree to calculate Faith's PD 
#this step is computationally expensive and needs high performance computing
seqs <- getSequences(masv); 
names(seqs) <- seqs;
alignment <- AlignSeqs(DNAStringSet(seqs), 
                       anchor=NA,verbose=FALSE);
phangAlign <- phyDat(as(alignment, "matrix"), type="DNA"); 
dm <- dist.ml(phangAlign); 
treeNJ <- NJ(dm); 
fit = pml(treeNJ, data=phangAlign);
fitGTR <- update(fit, k=4, inv=0.2); 
fitGTR <- optim.pml(fitGTR, model="GTR", optInv=TRUE, 
                    optGamma=TRUE, rearrangement = "stochastic", 
                    control = pml.control(trace = 0));

#save the output as .Rdata
save(fitGTR, file="data/ASV_phylo_tree_alphadiv.Rdata")